### PR TITLE
Support color ansi code sequences in custom help

### PR DESCRIPTION
**Pull request by shadowspawn**
# Pull Request

## Problem

People would like to add color to the Help: #301 (opened 10 years ago!).

Part of the intent of #1365 was to allow adding colour, but due to interactions with padding and wrapping it was too hard.

## Solution

Make it easy to add colour and styling. Treat ansi code sequences as zero width when padding and wrapping.

There is a new layer of `style` hooks which are just passed a string. There is not an assumption that only the style hooks do styling. So for example either of `optionTerm()` or `styleOptionTerm()` could add styling.

Build in support for common patterns for controlling colour (like `NO_COLOR`), and check output stream for color support. 

Strip escape sequences if necessary before displaying. So author does not need to check whether to use color when styling strings.

Refactor complex `wrap()` into more focused `formatItem()` and `boxWrap()`.

Now easy to do some previously messy help customisations (examples added):
- man-style help with item descriptions on new lines
- use `util:styleText` to add color to help
- use Chalk for color and padding and wrapping support and completely replace default implementation

### Changes from previous implementation

Wrapping of a very long single "word" is now a soft wrap rather than a hard wrap. Simpler implementation, and probably rare.

Dropped the partial wrapping support for Zero Width Space (ZWSP) as not displayed as zero width in Mac Terminal (among others), not consistent with general treatment of ansi escape sequences in this PR, and unlikely to be used by authors. For interest, another similar special character is the the [soft hyphen](https://en.wikipedia.org/wiki/Soft_hyphen). 

### Add

- support for adding custom color and style to Help output
- simple hooks for styling `Help` using `Command.configureHelp()`
    - `styleTitle()`
    - `styleUsage()` and `styleCommandText`
    - `styleItemDescription()`, `styleCommandDescription`(), and `styleDescriptionText()`
    - `styleOptionTerm`(), `styleSubcommandTerm()`, and `styleArgumentTerm()`
    - `styleOptionText()`, `styleSubcommandText()`, and `styleArgumentText()`
- new `Help` class methods to support color and allow easier customisation
    - `minWidthToWrap` property
    - `prepareContext()`
    - `displayWidth()`
    - `boxWrap()`
    - `formatItem()`
    - `preformatted()`
- `Command.configureOutput()` now includes `getOutHasColors()`, `getErrHasColours()`, and `stripColor()`

### Deleted

- `Help.wrap()` (refactored into `formatItem` and `boxWrap`)

### CODE REVIEW

**Comment by shadowspawn** [comment]
I added a whole bunch of simple style hooks. Some notes on current state of play.

I am not planning to add full JSDoc typing, keep the code compact. Most of them are one line of code.

The implementation of `styleUsage` is not super robust. But it does allow author to have matching colours for options and subcommands and arguments in the usage and the relevant section in help with just 3 simple style hooks.

It would reduce some boilerplate code if `optionTerm` called `styleOptionTerm` (and all the other similar pairs!), but I currently like the clean separation.

It is currently fine for authors to put color into the methods like `optionTerm`, they don't have to use `styleOptionTerm`.

The style routines get called even if color will ultimately be suppressed. The author does not need to know the rules. The current implementation strips colours later if color is not appropriate.

**Comment by shadowspawn** [comment]
There isn't an npm prerelease, but you can install it using git protocol (with npm).

For example, install from the current code on this pull request:

```
npm install git+https://github.com:tj/commander.js#pull/2251
```

**Comment by shadowspawn** [comment]
(I will rebase onto `release/13.x`.)

**Comment by manuel3108** [comment]
> There isn't an npm prerelease, but you can install it using git protocol (with npm).

Thanks, this is the actual url, in case anybody want's to give it a try.
```
npm install git+https://github.com:shadowspawn/commander.js#feature/colour-help
```

Overall this works very well, nice job.

File: lib/help.js [hunk 1]
```diff
@@ -471,50 +586,123 @@ class Help {
...
+ * Strip style ANSI escape sequences from the string. In particular, SGR (Select Graphic Rendition) codes.
+ *
+ * @param {string} str
+ * @returns {string}
+ * @package
+ */
+
+function stripColor(str) {
+  // eslint-disable-next-line no-control-regex
+  this.sgrPattern = /\x1b\[\d*(;\d*)*m/g;
```
[the remark this answers is not in the document]
Comment (shadowspawn) [reply lib/help.js:703]: No! Oops. I had been wondering about caching the regexp on the Help object and didn't tidy up when I moved the code out into a separate function.

**Review by abetomo — APPROVED** [review]
Awesome!

**Comment by shadowspawn** [comment]
Thanks @abetomo 🌈

**Comment by shadowspawn** [comment]
A prerelease is available for v13. The release is tagged as next and can be installed with:

    npm install commander@next

**Comment by shadowspawn** [comment]
Released in Commander v13.0.0

**Pull request merged**